homeassistant/components/binary_sensor/xiaomi.py

```diff
@@ -0,0 +1,316 @@
+        self._density = 0
+        XiaomiDevice.__init__(self, device, name, xiaomi_hub)
+
+    @property
+    def should_poll(self):
+        """Return True if entity has to be polled for state."""
+        return self._should_poll
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        return self._state
+
+    @property
+    def device_class(self):
+        """Return the class of binary sensor."""
+        return self._device_class
+
+    def update(self):
+        """Update the sensor state."""
+        _LOGGER.debug('Updating xiaomi sensor by polling')
+        self._get_from_hub(self._sid)
+
+
+class XiaomiNatgasSensor(XiaomiBinarySensor):
+    """Representation of a XiaomiNatgasSensor."""
+
+    def __init__(self, device, xiaomi_hub):
+        """Initialize the XiaomiSmokeSensor."""
+        self._data_key = 'alarm'
+        self._density = None
+        self._device_class = 'gas'
+        XiaomiBinarySensor.__init__(self, device, 'Natgas Sensor', xiaomi_hub)
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes."""
+        attrs = {ATTR_DENSITY: self._density}
+        attrs.update(super().device_state_attributes)
+        return attrs
+
+    def parse_data(self, data):
+        """Parse data sent by gateway."""
+        if DENSITY in data:
+            self._density = int(data.get(DENSITY))
+
+        value = data.get(self._data_key)
+        if value is None:
+            return False
+
+        if value == '1':
+            if self._state:
+                return False
+            self._state = True
+            return True
+        elif value == '0':
+            if self._state:
+                self._state = False
+                return True
+            return False
+
+
+class XiaomiMotionSensor(XiaomiBinarySensor):
+    """Representation of a XiaomiMotionSensor."""
+
+    def __init__(self, device, hass, xiaomi_hub):
+        """Initialize the XiaomiMotionSensor."""
+        self._hass = hass
+        self._data_key = 'status'
+        self._no_motion_since = 0
+        self._device_class = 'motion'
+        XiaomiBinarySensor.__init__(self, device, 'Motion Sensor', xiaomi_hub)
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes."""
+        attrs = {ATTR_NO_MOTION_SINCE: self._no_motion_since}
+        attrs.update(super().device_state_attributes)
+        return attrs
+
+    def parse_data(self, data):
+        """Parse data sent by gateway."""
+        self._should_poll = False
+        if NO_MOTION in data:  # handle push from the hub
+            self._no_motion_since = data[NO_MOTION]
+            self._state = False
+            return True
+
+        value = data.get(self._data_key)
+        if value is None:
+            return False
+
+        if value == MOTION:
+            self._should_poll = True
+            if self.entity_id is not None:
+                self._hass.bus.fire('motion', {
```

Comment:
HASS only fires when the sate changes.
This is cheap sensors, and sometimes the "no-motion" signal is not send. Then the HASS-sensor will continue to report motion. But this fires each time motion is registered, and will make automations more reliable.